redis/mutation_utils.cc

```diff
@@ -78,11 +102,63 @@ mutation make_mutation(service::storage_proxy& proxy, const redis_options& optio
     return m;
 }
 
-future<> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes&& key, bytes&& data, long ttl, service_permit permit) {
+future<bool> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes&& key, bytes&& data, long ttl, service_permit permit, smp_service_group& ssg) {
+    // construct cas_request
     db::timeout_clock::time_point timeout = db::timeout_clock::now() + options.get_write_timeout();
-    auto m = make_mutation(proxy, options, std::move(key), std::move(data), ttl);
+    auto schema = get_schema(proxy, options.get_keyspace_name(), redis::STRINGs);
+    const column_definition& column = *schema->get_column_definition(redis::DATA_COLUMN_NAME);
+    auto pkey = partition_key::from_single_value(*schema, key);
+    auto dk = dht::decorate_key(*schema, std::move(pkey));
+    auto partition_range = dht::partition_range::make_singular(dk);
+    dht::partition_range_vector partition_ranges;
+    partition_ranges.emplace_back(std::move(partition_range));
     auto write_consistency_level = options.get_write_consistency_level();
-    return proxy.mutate(std::vector<mutation> {std::move(m)}, write_consistency_level, timeout, nullptr, permit);
+    auto request = seastar::make_shared<write_strings_cas_request>(proxy, options, key, data, ttl);
+
+    // construct read command
+    auto ps = partition_slice_builder(*schema).build();
+    const auto max_result_size = proxy.get_max_result_size(ps);
+    query::read_command read_cmd(schema->id(), schema->version(), ps, 1, gc_clock::now(), std::nullopt, 1, utils::UUID(), query::is_first_page::no, max_result_size, 0);
+
+    service::client_state& client_state = options.get_client_state();
+    auto trace_state = tracing::trace_state_ptr();
+    auto desired_shard = service::storage_proxy::cas_shard(*schema, dk.token());
+    if (desired_shard == this_shard_id()) {
+        return proxy.cas(schema, request, make_lw_shared<query::read_command>(std::move(read_cmd)), partition_ranges,
+            {timeout, std::move(permit), client_state, trace_state},
+            db::consistency_level::LOCAL_SERIAL, db::consistency_level::LOCAL_QUORUM, timeout, timeout);
+    } else {
+        return proxy.container().invoke_on(desired_shard, ssg,
```

Comment:
If you rewrite it as a coroutine you won't need the do_with. For example, here you'd replace `return` with `co_return co_await`.

Reply:
(coroutines automatically copy variables to the heap so they have a stable address)